xwiki-commons-core/xwiki-commons-job/src/main/java/org/xwiki/job/internal/DefaultGroupedJobInitializerManager.java

```diff
@@ -0,0 +1,142 @@
+
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.xwiki.cache.Cache;
+import org.xwiki.cache.CacheException;
+import org.xwiki.cache.CacheManager;
+import org.xwiki.cache.config.LRUCacheConfiguration;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.job.GroupedJobInitializer;
+import org.xwiki.job.GroupedJobInitializerManager;
+import org.xwiki.job.JobGroupPath;
+import org.xwiki.job.JobManagerConfiguration;
+
+/**
+ * A component dedicated to find the appropriate {@link GroupedJobInitializer} based on a {@link JobGroupPath}.
+ *
+ * @since 12.5RC1
+ * @version $Id$
+ */
+@Component
+@Singleton
+public class DefaultGroupedJobInitializerManager implements GroupedJobInitializerManager, Initializable
+{
+    @Inject
+    @Named("context")
+    private Provider<ComponentManager> componentManagerProvider;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private GroupedJobInitializer defaultGoupedJobInitializer;
+
+    @Inject
+    private CacheManager cacheManager;
+
+    @Inject
+    private JobManagerConfiguration configuration;
+
+    private Cache<GroupedJobInitializer> cachedInitializers;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        LRUCacheConfiguration cacheConfiguration = new LRUCacheConfiguration("job.grouped.initializer",
+            this.configuration.getGroupedJobInitializerCacheSize());
+        try {
+            this.cachedInitializers = this.cacheManager.createNewCache(cacheConfiguration);
+        } catch (CacheException e) {
+            throw new InitializationException("Error while creating cache for GroupedJobInitializer", e);
+        }
+    }
+
+    /**
+     * Search a {@link GroupedJobInitializer} by checking if their ID matches the given {@link JobGroupPath}.
+     * This methods fills the cache with initializer information as side effect.
+     * This methods performs the following search:
+     *   1. check if the initializer is already stored in the cache
+     *   2. if not stored, iterates over all the {@link GroupedJobInitializer}, put them in cache, and check if it
+     *      matches the path. If it matches stop iterating.
+     * @param path the searched path.
+     * @return an initializer matching the path or null if no initializer has been found.
+     */
+    private GroupedJobInitializer search(JobGroupPath path)
+    {
+        GroupedJobInitializer result = this.cachedInitializers.get(path.toString());
+        if (result == null) {
+            try {
+                List<GroupedJobInitializer> initializers =
```

Comment:
Wouldn't it make more sense from performance point of view to give this list to `search()` instead of recreating it for each element of the path ? The list is always the same during the `for` in getGroupedJobInitializer.

Reply:
Might make sense since we don't trust much the cache and we are always looping on the components to find an exact match which might be the exception: more likely we'll fallback on a parent initializer so we'll loop there at least once but maybe twice or more.